libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,611 @@
+        timestamp = float(time_week_ms % (3600 * 1000)) * 0.001;
+        timestamp = create_location_timestamp(timestamp);   //make sure timestamp is within Remote ID limit
+    }
+
+
+    {
+        WITH_SEMAPHORE(_sem);
+        // take semaphore so CAN gets a consistent packet
+        pkt_location = {
+        latitude : latitude,
+        longitude : longitude,
+        altitude_barometric : altitude_barometric,
+        altitude_geodetic : altitude_geodetic,
+        height : height_above_takeoff,
+        timestamp : timestamp,
+        direction : uint16_t(direction * 100.0), // Heading (centi-degrees)
+        speed_horizontal : uint16_t(speed_horizontal * 100.0), // Ground speed (cm/s)
+        speed_vertical : int16_t(climb_rate * 100.0), // Climb rate (cm/s)
+        target_system : 0,
+        target_component : 0,
+        status : uint8_t(uav_status),
+        height_reference : MAV_ODID_HEIGHT_REF_OVER_TAKEOFF,           // height reference enum: Above takeoff location or above ground
+        horizontal_accuracy : uint8_t(horizontal_accuracy_mav),
+        vertical_accuracy : uint8_t(vertical_accuracy_mav),
+        barometer_accuracy : barometer_accuracy,
+        speed_accuracy : uint8_t(speed_accuracy_mav),
+        timestamp_accuracy : uint8_t(timestamp_accuracy_mav)
+        };
+        need_send_location = dronecan_send_all;
+    }
+
+    if (_chan != MAV_CHAN_INVALID) {
+        mavlink_msg_open_drone_id_location_send_struct(_chan, &pkt_location);
+    }
+}
+
+void AP_OpenDroneID::send_basic_id_message()
+{
+    // note that packet is filled in by the GCS
+    need_send_basic_id |= dronecan_send_all;
+    if (_chan != MAV_CHAN_INVALID) {
+        mavlink_msg_open_drone_id_basic_id_send_struct(_chan, &pkt_basic_id);
+    }
+}
+
+void AP_OpenDroneID::send_system_message()
+{
+    // note that packet is filled in by the GCS
+    need_send_system |= dronecan_send_all;
+    if (_chan != MAV_CHAN_INVALID) {
+        mavlink_msg_open_drone_id_system_send_struct(_chan, &pkt_system);
+    }
+}
+
+void AP_OpenDroneID::send_self_id_message()
+{
+    need_send_self_id |= dronecan_send_all;
+    // note that packet is filled in by the GCS
+    if (_chan != MAV_CHAN_INVALID) {
+        mavlink_msg_open_drone_id_self_id_send_struct(_chan, &pkt_self_id);
+    }
+}
+
+void AP_OpenDroneID::send_operator_id_message()
+{
+    need_send_operator_id |= dronecan_send_all;
+    // note that packet is filled in by the GCS
+    if (_chan != MAV_CHAN_INVALID) {
+        mavlink_msg_open_drone_id_operator_id_send_struct(_chan, &pkt_operator_id);
+    }
+}
+
+/*
+* This converts a horizontal accuracy float value to the corresponding enum
+*
+* @param Accuracy The horizontal accuracy in meters
+* @return Enum value representing the accuracy
+*/
+MAV_ODID_HOR_ACC AP_OpenDroneID::create_enum_horizontal_accuracy(float accuracy) const
+{
+    // Out of bounds return UKNOWN flag
+    if (accuracy < 0.0 || accuracy >= 18520.0) {
```

Comment:
Why bother checking the maximum here? It just means we have to keep that updated with table. We need to sanity check the low side obviously, but the final return statement will cover the high side for us nicely.